src/main/drivers/bus_i2c.c

```diff
@@ -0,0 +1,441 @@
+﻿/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <platform.h>
+#include "build_config.h"
+
+#include "gpio.h"
+#include "system.h"
+
+#include "bus_i2c.h"
+#include "nvic.h"
+#include "drivers/io_impl.h"
+#include "drivers/rcc.h"
+
+#ifndef SOFT_I2C
+
+static void i2c_er_handler(I2CDevice device);
+static void i2c_ev_handler(I2CDevice device);
+static void i2cUnstick(IO_t scl, IO_t sda);
+
+#define GPIO_AF_I2C GPIO_AF_I2C1
+
+#if defined(STM32F40_41xxx) || defined(STM32F411xE)
+#ifndef I2C1_SCL 
+#define I2C1_SCL PB8 
+#endif
+#ifndef I2C1_SDA 
+#define I2C1_SDA PB9 
+#endif
+#else
+#ifndef I2C1_SCL 
+#define I2C1_SCL PB6 
+#endif
+#ifndef I2C1_SDA 
+#define I2C1_SDA PB7 
+#endif
+#endif
+
+#ifndef I2C2_SCL 
+#define I2C2_SCL PB10 
+#endif
+#ifndef I2C2_SDA 
+#define I2C2_SDA PB11
+#endif
+
+#if defined(STM32F40_41xxx) || defined(STM32F411xE)
+#ifndef I2C3_SCL 
+#define I2C3_SCL PA8
+#endif
+#ifndef I2C3_SDA 
+#define I2C3_SDA PB4 
+#endif
+#endif
+
+static i2cDevice_t i2cHardwareMap[] = {
+    { I2C1, IO_TAG(I2C1_SCL), IO_TAG(I2C1_SDA), RCC_APB1(I2C1), false, I2C1_EV_IRQn, I2C1_ER_IRQn },
```

Comment:
So that you can use single I2C on different pins concurrently? Will work, but sounds a bit too complicated.

Or select one configuration on runtime? Just add array of possible pinTag+AF configurations (limited to some reasonable number on F3).

If fully general IO configuration is supported, then it would be probably best to tag IO pin with function:
`PA3 - I2C(1)` and then search possible configurations to get correct AF (this is important on F3 ADC - single pin is connected to different ADCs, it is not easy to select 'best' one)